feat: keep cached user stats [BD-38]

xitij2000 opened the pull request:
In a previous PR (https://github.com/edx/cs_comments_service/pull/352) we introduced a new API that would generate user stats for a course on-the-fly. The query would return data about all the users in a course who'd participated in discussions. In some cases, this could be in hundreds of thousands or millions of users, and result in dozens if not hundreds of MBs of data. 

This data also needed to be sorted, and paginated. Due to the nature of the query, paginating this data would not reduce the performance implications of the query, since it'd still need to query all the course data to build the stats. That said, it would reduce the final amount of data involved.

This PR implements a different mechanism for the same. It will keep track of user counts across courses, by incrementing the counts each time a user creates a post, comment etc. It will also track deletes, and when content is marked/unmarked as abusive.  The first time such an increment/decrement happens it will auto-backfill the data to keep counts accurate. However, other than that it will not automatically backfill data for users yet.

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [BLENDED-1044](https://openedx.atlassian.net/browse/BLENDED-1044) to keep track of it in Jira.   More details are on the [BD-38](https://openedx.atlassian.net/wiki/spaces/COMM/pages/2790490201/BD-38+Discussions+In-Context+Discussion+Moderation+Phase+2) project page.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

xitij2000 commented on api/users.rb:
```diff
@@ -11,11 +11,54 @@
   end
 end
 
+get "#{APIPREFIX}/users/:course_id/stats2" do |course_id|
```
This is a temporary name to allow testing both endpoints. This will replace stats.

xitij2000 commented on spec/api/user_spec.rb:
```diff
@@ -428,11 +427,13 @@ def add_flags(content, expected_result)
         end
+        # Sort the map entries using the default sort
+        expected_result = expected_result.values.sort_by  { |val| [val["threads"], val["responses"], val["replies"]] }.reverse
 
-        get "/api/v1/users/#{course_id}/stats"
+        get "/api/v1/users/#{course_id}/stats2"
         expect(last_response.status).to eq(200)
         res = parse(last_response.body)
-        expect(res).to eq expected_result
+        expect(res["user_stats"]).to eq expected_result
```
The new API now returns a list of objects sorted by activity but other than that the results should be the same as before. 

I'll add more tests for checking that adding comments, removing comments, etc keep the counts updated.

xitij2000 commented:
@mtyaka Could you do a review of this PR?

xitij2000 commented on api/users.rb:
```diff
@@ -11,6 +11,50 @@
+                      .paginate(:page => page, :per_page => per_page)
+  total_count = paginated_stats.total_entries
+
+  data = paginated_stats.to_a.map { |user_stats| {
+    :username => user_stats["username"]
+  }.merge(
+    user_stats["course_stats"].first.except("_id", "course_id")
+  ) }
```
Flatten the nesting, and remove the ids from each document since the id is unneeded, and the course_id is common.

xitij2000 commented on api/users.rb:
```diff
@@ -11,6 +11,50 @@
+
+  paginated_stats = User
+                      .where("course_stats.course_id" => course_id)
+                      .only(:username, :'course_stats.$')
```
Only return the username field, and the `course_stats` entry that matches the `course_id` condition.

xitij2000 commented on models/comment.rb:
```diff
@@ -185,3 +204,63 @@ def set_sk
+
+
+def build_course_stats_for_user(user, course_id)
+  data = Content.collection.aggregate(
+    [
+      # Match all content in the course
+      { "$match" => { :course_id => course_id, :author_id => user.external_id } },
+      # Keep a count of flags for each entry
+      {
+        "$set" => {
+          # Just using $ne with null will return true if the field is absent
+          # So we first fall all absent fields to null, then check if it's null,
+          # that way we match for the absence of the field or value = null
+          :is_reply => { "$ne" => [{ "$ifNull" => ["$parent_id", nil] }, nil] }
+        }
+      },
+      {
+        "$group" => {
+          # Here we're grouping items by the type (comment or thread), and the user, and whether the comment is a reply.
+          # For threads is_reply will always be false.
+          :_id => { :type => "$_type", :is_reply => "$is_reply" },
+          # This will just count each group, so we get a breakdown of how many comments and threads a user has created.
+          :count => { "$sum" => 1 },
+          # These two will sum up the active and inactive reports in each category
+          # i.e. reported threads, reported comments, reported replies
+          # The way this works is (starting from inside out), we take the size of the abuse_flaggers list, we compare
+          # it to 0 using $cmp. If it's greater than zero then $cmp results in 1 otherwise 0.
+          # So we're summing up 1 for each abuse_flagger array that has entries, and 0 for the rest. This gives us a
+          # count of
+          :active_flags => { "$sum" => { "$cmp" => [{ "$size" => "$abuse_flaggers" }, 0] } },
+          :inactive_flags => { "$sum" => { "$cmp" => [{ "$size" => "$historical_abuse_flaggers" }, 0] } },
+        }
+      }
+    ])
```
We use the same query, but scoped to a single user, so it's a much smaller set of data to go through.

mtyaka commented on spec/api/user_spec.rb:
```diff
@@ -410,29 +411,149 @@ def add_flags(content, expected_result)
         end
+        expected_data
+      end
+
```
Nit: extra newline

mtyaka commented on spec/api/user_spec.rb:
```diff
@@ -410,29 +411,149 @@ def add_flags(content, expected_result)
+        expect(new_stats["threads"]).to eq original_stats["threads"]
+        expect(new_stats["responses"]).to eq original_stats["responses"]
+        expect(new_stats["replies"]).to eq original_stats["replies"] -1
+        end
```
nit: indentation

mtyaka reviewed (changes requested):
The code looks good @xitij2000! I posted a few comments, but they are mostly nits. The only larger concern is that it looks like we're treating empty arrays as falsey in some places, but empty arrays are truthy in ruby.

I didn't do any manual testnig. Do we have a sandbox where I could test it out?

xitij2000 commented on models/comment.rb:
```diff
@@ -66,6 +66,25 @@ def as_indexed_json(options={})
+      end
+    end
+  end
+  after_save do
```
Makes sense! I don't think the author can change, and I guess this will track each edit.

xitij2000 commented on models/user.rb:
```diff
@@ -232,14 +233,41 @@ def replace_username_in_all_content(new_username)
     read_state.save
   end
 
+  def stats_for_course(course_id)
```
Sure, I think the place where it was used got refactored out, will remove.

mtyaka commented on lib/helpers.rb:
```diff
@@ -55,7 +55,7 @@ def vote_for(obj)
     raise ArgumentError, t(:user_id_is_required) unless user
     # If this is the first abuse flag, count it, otherwise don't increment the count
-    obj.author.update_stats_for_course(obj.course_id, active_flags: 1) unless obj.abuse_flaggers
+    obj.author.update_stats_for_course(obj.course_id, active_flags: 1) unless obj.abuse_flaggers.empty?
```
This should be `if obj.abuse_flaggers.empty?` instead of `unless`, if I understand correctly.

mtyaka commented on lib/helpers.rb:
```diff
@@ -65,15 +65,15 @@ def un_flag_as_abuse(obj)
     if params["all"]
       # If historical abuse flags didn't exist so far, increment them.
-      obj.author.update_stats_for_course(obj.course_id, inactive_flags: 1) unless obj.historical_abuse_flaggers
+      obj.author.update_stats_for_course(obj.course_id, inactive_flags: 1) unless obj.historical_abuse_flaggers.empty?
```
This should be `if obj.abuse_flaggers.empty?` instead of `unless`, if I understand correctly.

mtyaka commented on lib/helpers.rb:
```diff
@@ -65,15 +65,15 @@ def un_flag_as_abuse(obj)
     end
     # If this operation results in no more abuse flags, decrement the cour.
-    obj.author.update_stats_for_course(obj.course_id, active_flags: -1) unless obj.abuse_flaggers
+    obj.author.update_stats_for_course(obj.course_id, active_flags: -1) unless obj.abuse_flaggers.empty?
```
This should be `if obj.abuse_flaggers.empty?` instead of `unless`, if I understand correctly.

mtyaka reviewed (changes requested):
The code looks great @xitij2000! The three `unless` have to be changed to `if`s, which I think is the reason why tests are failing.

Do we have a sandbox where I could test this?

mtyaka commented on lib/helpers.rb:
```diff
@@ -54,6 +54,8 @@ def vote_for(obj)
   def flag_as_abuse(obj)
     raise ArgumentError, t(:user_id_is_required) unless user
+    # If this is the first abuse flag, count it, otherwise don't increment the count
+    obj.author.update_stats_for_course(obj.course_id, active_flags: 1) if obj.abuse_flaggers.empty?
```
Sorry for not noticing this earlier, but I think there's a subtle bug here -- if stats for this course don't exist yet, `update_stats_for_course()` will automatically generate them from existing data, but when doing that, it will not see the update to `obj.abuse_flaggers`.

So in case:
1. course stats don't yet exist; and
2. this is the first abuse flag for this object

Then course stats will be automatically generated from existing data and will not see the abuse flag on this comment.

mtyaka replied:
A similar thing can happen inside the `un_flag_as_abuse` method as well.

xitij2000 replied:
Ah! That explains it! I was looking into this exact issue, but didn't have much time to work on friday. I'll look into a fix for this.

mtyaka reviewed (commented):
@xitij2000 I found a subtle bug in the `flag_as_abuse` and `un_flag_as_abuse` methods. 

However those don't explain the failing tests related to this method. The errors look really baffling, I've been trying to figure out why they happen without luck :/

mtyaka commented:
> However those don't explain the failing tests related to this method. The errors look really baffling, I've been trying to figure out why they happen without luck :/

Actually the first two failures might be related to that, but the "handles removing flags" error makes no sense to me.

xitij2000 commented:
> Actually the first two failures might be related to that, but the "handles removing flags" error makes no sense to me.

I tried isolating changes since the last passing tests and it seems that the errors are due to a change from after_save to after_create. Not sure why, but looking into it.

xitij2000 commented:
> I tried isolating changes since the last passing tests and it seems that the errors are due to a change from after_save to after_create. Not sure why, but looking into it.

OK, I figured it out. When using after_save, the stats are updated after the flags are added in the test data, however since the flags are added without triggering the update mechanism they are not updated when you switch to after_create. 

I've now updated the tests so that we run an initial update of course stats, and then perform the rest of the tests. This fixes teh issues.

mtyaka commented:
> When using after_save, the stats are updated after the flags are added in the test data, however since the flags are added without triggering the update mechanism they are not updated when you switch to after_create.

Ah, that makes sense.

Code looks good to me 👍

felipetrz commented on models/user.rb:
```diff
@@ -232,14 +233,32 @@ def replace_username_in_all_content(new_username)
+      stats.inc(responses: responses) if responses
+      stats.inc(replies: replies) if replies
+      stats.inc(active_flags: active_flags) if active_flags
+      stats.inc(inactive_flags: inactive_flags) if inactive_flags
```
This is doing multiple roundtrips to mongodb every time multiple fields need to be updated.
Instead of defaulting to `nil`, this could simply default to `0` and pass all the parameters into a single `inc` call.

xitij2000 replied:
@felipetrz It won't since we never increment more than one item at a time. At any point only one of these parameters will be passed, so only one of these values will be incremented. 

That said, while I did this to keep the function explicit, I'll just simplify it so it's easier to follow.

felipetrz commented on models/user.rb:
```diff
@@ -232,14 +233,32 @@ def replace_username_in_all_content(new_username)
+      stats.inc(inactive_flags: inactive_flags) if inactive_flags
+    rescue Mongoid::Errors::DocumentNotFound
+      # If the stats don't already exist, rebuild them from scratch, in which case they will have the correct counts.
+      build_course_stats_for_user(self, course_id)
```
Since this rebuild happens as a trigger from posting/deleting comments/threads, can't this give unexpected slow responses, or even timeouts, the first time the stats are generated in big courses?
Maybe it'd be better to pre-generate stats as a migration step, instead of sneaking it into the `after_*` hooks.

felipetrz replied:
Oh, I just saw the comment about this being scoped to the user, so this should be fine. Nevermind this.

xitij2000 replied:
I did consider making this a background task, however, for a single user and course, it should be a pretty small query, even for a very active user.

felipetrz reviewed (changes requested):
I commented on a couple potential performance concerns still left, but otherwise this is looking a lot better than the previous solution.

mtyaka commented:
Nice work @xitij2000! :+1:

- [x] I tested this on the https://discussions.sandbox.opencraft.hosting/ sandbox. I made a few posts, added a few replies, and flaged/unflagged a few other users' posts. I verified that [the stats API endpoint](https://discussions.sandbox.opencraft.hosting/api/discussion/v1/courses/course-v1:Infinity+Math101+2021_Winter/activity_stats?order_by=activity) returned correct results after each change.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ N/A
- [x] Includes documentation -- includes inline comments.

edx-webhook commented:
@xitij2000 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/cs_comments_service&entry.867055334=https://github.com/openedx/cs_comments_service/pull/361&entry.1484655318=https://github.com/xitij2000&entry.752974735=2021-12-10+13:41&entry.1917517419=2022-01-25+12:45&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
